File: Go/crypto_utils/crypto_utils_test.go
package crypto_utils

import (
	"strings"
	"testing"
)

// ============================================================================
// Hash Function Tests
// ============================================================================

func TestMd5Hash(t *testing.T) {
	tests := []struct {
		input    string
		expected string
	}{
		{"hello", "5d41402abc4b2a76b9719d911017c592"},
		{"", "d41d8cd98f00b204e9800998ecf8427e"},
		{"Hello, World!", "65a8e27d8879283831b664bd8b7f0ad4"},
		{"中文测试", "a7fc8e26e9be5b1b1c42f5a1c48c25d1"},
	}

	for _, test := range tests {
		result := Md5Hash(test.input)
		if result != test.expected {
			t.Errorf("Md5Hash(%q) = %q, expected %q", test.input, result, test.expected)
		}
	}
}

func TestSha1Hash(t *testing.T) {
	tests := []struct {
		input    string
		expected string
	}{
		{"hello", "aaf4c61ddcc5e8a2dabede0f3b482cd9aea9434d"},
		{"", "da39a3ee5e6b4b0d3255bfef95601890afd80709"},
		{"Hello, World!", "0a0a9f2a6772942557ab5355d76af442f8f65e01"},
	}

	for _, test := range tests {
		result := Sha1Hash(test.input)
		if result != test.expected {
			t.Errorf("Sha1Hash(%q) = %q, expected %q", test.input, result, test.expected)
		}
	}
}

func TestSha256Hash(t *testing.T) {
	tests := []struct {
		input    string
		expected string
	}{
		{"hello", "2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824"},
		{"", "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"},
	}

	for _, test := range tests {
		result := Sha256Hash(test.input)
		if result != test.expected {
			t.Errorf("Sha256Hash(%q) = %q, expected %q", test.input, result, test.expected)
		}
	}
}

func TestSha512Hash(t *testing.T) {
	result := Sha512Hash("hello")
	if len(result) != 128 {
		t.Errorf("Sha512Hash length = %d, expected 128", len(result))
	}
}

func TestSha256HashBytes(t *testing.T) {
	data := []byte("hello")
	hash := Sha256HashBytes(data)
	if len(hash) != 32 {
		t.Errorf("Sha256HashBytes length = %d, expected 32", len(hash))
	}
}

// ============================================================================
// HMAC Function Tests
// ============================================================================

func TestHmacSha256(t *testing.T) {
	result := HmacSha256("hello", "secret")
	if len(result) != 64 {
		t.Errorf("HmacSha256 length = %d, expected 64", len(result))
	}

	// Same input should produce same output
	result2 := HmacSha256("hello", "secret")
	if result != result2 {
		t.Error("HmacSha256 should produce consistent results")
	}

	// Different key should produce different output
	result3 := HmacSha256("hello", "different")
	if result == result3 {
		t.Error("HmacSha256 with different keys should produce different results")
	}
}

// ============================================================================
// Base64 Function Tests
// ============================================================================

func TestBase64Encode(t *testing.T) {
	tests := []struct {
		input    string
		expected string
	}{
		{"hello", "aGVsbG8="},
		{"", ""},
		{"Hello, World!", "SGVsbG8sIFdvcmxkIQ=="},
	}

	for _, test := range tests {
		result := Base64Encode(test.input)
		if result != test.expected {
			t.Errorf("Base64Encode(%q) = %q, expected %q", test.input, result, test.expected)
		}
	}
}

func TestBase64Decode(t *testing.T) {
	tests := []struct {
		input    string
		expected string
	}{
		{"aGVsbG8=", "hello"},
		{"", ""},
		{"SGVsbG8sIFdvcmxkIQ==", "Hello, World!"},
	}

	for _, test := range tests {
		result := Base64Decode(test.input)
		if result != test.expected {
			t.Errorf("Base64Decode(%q) = %q, expected %q", test.input, result, test.expected)
		}
	}
}

func TestBase64RoundTrip(t *testing.T) {
	tests := []string{
		"hello",
		"Hello, World!",
		"中文测试",
		"Special chars: !@#$%",
		"",
	}

	for _, test := range tests {
		encoded := Base64Encode(test)
		decoded := Base64Decode(encoded)
		if decoded != test {
			t.Errorf("Base64 round-trip failed for %q: got %q", test, decoded)
		}
	}
}

func TestBase64UrlEncode(t *testing.T) {
	// Test that URL-safe encoding doesn't contain + or /
	input := "hello+world/test"
	encoded := Base64UrlEncode(input)
	if strings.Contains(encoded, "+") || strings.Contains(encoded, "/") {
		t.Error("Base64UrlEncode should not contain + or /")
	}
	if strings.Contains(encoded, "=") {
		t.Error("Base64UrlEncode should not contain padding =")
	}
}

func TestBase64UrlRoundTrip(t *testing.T) {
	tests := []string{
		"hello+world/test",
		"special>chars<here",
		"path/to/file.txt",
	}

	for _, test := range tests {
		encoded := Base64UrlEncode(test)
		decoded := Base64UrlDecode(encoded)
		if decoded != test {
			t.Errorf("Base64Url round-trip failed for %q: got %q", test, decoded)
		}
	}
}

func TestIsValidBase64(t *testing.T) {
	valid := []string{
		"aGVsbG8=",
		"SGVsbG8sIFdvcmxkIQ==",
		"",
	}
	invalid := []string{
		"not-valid!!!",
		"aGVsbG8",
	}

	for _, test := range valid {
		if !IsValidBase64(test) {
			t.Errorf("IsValidBase64(%q) should be true", test)
		}
	}

	for _, test := range invalid {
		if IsValidBase64(test) {
			t.Errorf("IsValidBase64(%q) should be false", test)
		}
	}
}

// ============================================================================
// Random Function Tests
// ============================================================================

func TestRandomString(t *testing.T) {
	// Test length
	for _, length := range []int{1, 8, 16, 32, 100} {
		result := RandomString(length, "")
		if len(result) != length {
			t.Errorf("RandomString(%d) length = %d, expected %d", length, len(result), length)
		}
	}

	// Test zero length
	result := RandomString(0, "")
	if result != "" {
		t.Error("RandomString(0) should return empty string")
	}

	// Test negative length
	result = RandomString(-1, "")
	if result != "" {
		t.Error("RandomString(-1) should return empty string")
	}

	// Test custom charset
	custom := "ABC"
	result = RandomString(10, custom)
	for _, c := range result {
		if c != 'A' && c != 'B' && c != 'C' {
			t.Error("RandomString with custom charset should only use provided characters")
		}
	}
}

func TestRandomStringUniqueness(t *testing.T) {
	// Generate multiple strings and check they're different
	seen := make(map[string]bool)
	for i := 0; i < 100; i++ {
		s := RandomString(16, "")
		if seen[s] {
			t.Error("RandomString should produce unique values")
			break
		}
		seen[s] = true
	}
}

func TestRandomPassword(t *testing.T) {
	// Test minimum length
	password := RandomPassword(2)
	if len(password) != 4 {
		t.Errorf("RandomPassword(2) length = %d, expected 4 (minimum)", len(password))
	}

	// Test specified length
	password = RandomPassword(16)
	if len(password) != 16 {
		t.Errorf("RandomPassword(16) length = %d, expected 16", len(password))
	}

	// Test character variety
	hasLower := false
	hasUpper := false
	hasDigit := false
	hasSpecial := false

	for _, c := range password {
		if c >= 'a' && c <= 'z' {
			hasLower = true
		} else if c >= 'A' && c <= 'Z' {
			hasUpper = true
		} else if c >= '0' && c <= '9' {
			hasDigit = true
		} else {
			hasSpecial = true
		}
	}

	if !hasLower {
		t.Error("RandomPassword should contain at least one lowercase letter")
	}
	if !hasUpper {
		t.Error("RandomPassword should contain at least one uppercase letter")
	}
	if !hasDigit {
		t.Error("RandomPassword should contain at least one digit")
	}
	if !hasSpecial {
		t.Error("RandomPassword should contain at least one special character")
	}
}

// ============================================================================
// UUID Function Tests
// ============================================================================

func TestGenerateUUID(t *testing.T) {
	uuid := GenerateUUID()

	// Check length
	if len(uuid) != 36 {
		t.Errorf("GenerateUUID length = %d, expected 36", len(uuid))
	}

	// Check format (8-4-4-4-12)
	if uuid[8] != '-' || uuid[13] != '-' || uuid[18] != '-' || uuid[23] != '-' {
		t.Error("GenerateUUID should have correct format with hyphens at positions 8, 13, 18, 23")
	}

	// Check version (4th segment should start with '4')
	if uuid[14] != '4' {
		t.Errorf("GenerateUUID version should be 4, got %c", uuid[14])
	}

	// Check variant (3rd segment should start with 8, 9, a, or b)
	variant := uuid[19]
	if variant != '8' && variant != '9' && variant != 'a' && variant != 'b' {
		t.Errorf("GenerateUUID variant should be 8, 9, a, or b, got %c", variant)
	}
}

func TestGenerateUUIDUniqueness(t *testing.T) {
	seen := make(map[string]bool)
	for i := 0; i < 100; i++ {
		uuid := GenerateUUID()
		if seen[uuid] {
			t.Error("GenerateUUID should produce unique values")
			break
		}
		seen[uuid] = true
	}
}

func TestGenerateUUIDSimple(t *testing.T) {
	uuid := GenerateUUIDSimple()

	// Check length (no hyphens)
	if len(uuid) != 32 {
		t.Errorf("GenerateUUIDSimple length = %d, expected 32", len(uuid))
	}

	// Check no hyphens
	if strings.Contains(uuid, "-") {
		t.Error("GenerateUUIDSimple should not contain hyphens")
	}
}

func TestIsValidUUID(t *testing.T) {
	validUUIDs := []string{
		"550e8400-e29b-41d4-a716-446655440000",
		"6ba7b810-9dad-11d1-80b4-00c04fd430c8",
		"00000000-0000-4000-8000-000000000000",
	}

	invalidUUIDs := []string{
		"not-a-uuid",
		"550e8400-e29b-11d4-a716-44665544000", // Too short
		"550e8400-e29b-11d4-a716-4466554400000", // Too long
		"550e8400_e29b_11d4_a716_446655440000", // Wrong separators
		"550e8400-e29b-11d4-g716-446655440000", // Invalid character 'g'
	}

	for _, uuid := range validUUIDs {
		if !IsValidUUID(uuid) {
			t.Errorf("IsValidUUID(%q) should be true", uuid)
		}
	}

	for _, uuid := range invalidUUIDs {
		if IsValidUUID(uuid) {
			t.Errorf("IsValidUUID(%q) should be false", uuid)
		}
	}
}

// ============================================================================
// XOR Function Tests
// ============================================================================

func TestXorEncryptDecrypt(t *testing.T) {
	tests := []struct {
		input string
		key   string
	}{
		{"hello", "secret"},
		{"Hello, World!", "mykey"},
		{"中文测试", "key"},
		{"", "key"},
		{"test", ""},
	}

	for _, test := range tests {
		encrypted := XorEncrypt(test.input, test.key)
		decrypted := XorDecrypt(encrypted, test.key)
		if decrypted != test.input {
			t.Errorf("XOR round-trip failed for %q with key %q: got %q", test.input, test.key, decrypted)
		}
	}
}

func TestXorDifferentKeys(t *testing.T) {
	input := "hello"
	key1 := "secret"
	key2 := "different"

	encrypted1 := XorEncrypt(input, key1)
	encrypted2 := XorEncrypt(input, key2)

	if encrypted1 == encrypted2 {
		t.Error("XOR encryption with different keys should produce different results")
	}
}

// ============================================================================
// Validation Function Tests
// ============================================================================

func TestIsValidHash(t *testing.T) {
	tests := []struct {
		hash      string
		algorithm string
		expected  bool
	}{
		{"5d41402abc4b2a76b9719d911017c592", "md5", true},
		{"5d41402abc4b2a76b9719d911017c59", "md5", false}, // Too short
		{"5d41402abc4b2a76b9719d911017c59g", "md5", false}, // Invalid char
		{"aaf4c61ddcc5e8a2dabede0f3b482cd9aea9434d", "sha1", true},
		{"2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824", "sha256", true},
		{"invalid", "sha256", false},
		{"valid", "unknown", false},
	}

	for _, test := range tests {
		result := IsValidHash(test.hash, test.algorithm)
		if result != test.expected {
			t.Errorf("IsValidHash(%q, %q) = %v, expected %v", test.hash, test.algorithm, result, test.expected)
		}
	}
}

func TestIsValidMd5(t *testing.T) {
	if !IsValidMd5("5d41402abc4b2a76b9719d911017c592") {
		t.Error("IsValidMd5 should return true for valid MD5")
	}
	if IsValidMd5("invalid") {
		t.Error("IsValidMd5 should return false for invalid MD5")
	}
}

func TestIsValidSha256(t *testing.T) {
	if !IsValidSha256("2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824") {
		t.Error("IsValidSha256 should return true for valid SHA256")
	}
	if IsValidSha256("invalid") {
		t.Error("IsValidSha256 should return false for invalid SHA256")
	}
}

// ============================================================================
// Constant Tests
// ============================================================================

func TestConstants(t *testing.T) {
	if len(LowerCaseLetters) != 26 {
		t.Error("LowerCaseLetters should have 26 characters")
	}
	if len(UpperCaseLetters) != 26 {
		t.Error("UpperCaseLetters should have 26 characters")
	}
	if len(Digits) != 10 {
		t.Error("Digits should have 10 characters")
	}
	if len(Alphanumeric) != 62 {
		t.Error("Alphanumeric should have 62 characters")
	}
}

File: VB/crypto_utils/crypto_utils_test.vb
' =============================================================================
' AllToolkit - Crypto Utilities Tests for VB.NET
' =============================================================================
' CryptoUtils 模块的单元测试
'
' 运行方式:
'   1. 使用 Visual Studio: 打开测试资源管理器并运行
'   2. 使用命令行: dotnet test (如果配置了 .NET SDK)
'   3. 使用 VSTest: vstest.console.exe crypto_utils_test.dll
'
' 作者: AllToolkit Contributors
' 许可证: MIT
' =============================================================================

Imports System
Imports System.IO
Imports System.Text
Imports Microsoft.VisualStudio.TestTools.UnitTesting
Imports AllToolkit

Namespace AllToolkit.Tests

    ''' <summary>
    ''' CryptoUtils 模块的单元测试类。
    ''' </summary>
    <TestClass>
    Public Class CryptoUtilsTest

        ' =====================================================================
        ' MD5 哈希测试
        ' =====================================================================

        <TestMethod>
        Public Sub TestMd5Hash()
            ' 已知 MD5 值测试
            Assert.AreEqual("5d41402abc4b2a76b9719d911017c592", CryptoUtils.Md5Hash("hello"))
            Assert.AreEqual("d41d8cd98f00b204e9800998ecf8427e", CryptoUtils.Md5Hash(""))
            
            ' 空值测试
            Assert.AreEqual("", CryptoUtils.Md5Hash(Nothing))
            
            ' 中文测试
            Assert.AreEqual(32, CryptoUtils.Md5Hash("你好世界").Length)
        End Sub

        <TestMethod>
        Public Sub TestMd5HashConsistency()
            ' 一致性测试
            Dim hash1 As String = CryptoUtils.Md5Hash("test123")
            Dim hash2 As String = CryptoUtils.Md5Hash("test123")
            Assert.AreEqual(hash1, hash2)
        End Sub

        ' =====================================================================
        ' SHA 哈希测试
        ' =====================================================================

        <TestMethod>
        Public Sub TestSha1Hash()
            Assert.AreEqual(40, CryptoUtils.Sha1Hash("hello").Length)
            Assert.AreEqual("", CryptoUtils.Sha1Hash(Nothing))
            Assert.AreEqual("", CryptoUtils.Sha1Hash(""))
        End Sub

        <TestMethod>
        Public Sub TestSha256Hash()
            Assert.AreEqual(64, CryptoUtils.Sha256Hash("hello").Length)
            Assert.AreEqual("", CryptoUtils.Sha256Hash(Nothing))
            Assert.AreEqual("", CryptoUtils.Sha256Hash(""))
            
            ' 已知 SHA256 值测试
            Dim expected As String = "2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824"
            Assert.AreEqual(expected, CryptoUtils.Sha256Hash("hello"))
        End Sub

        <TestMethod>
        Public Sub TestSha512Hash()
            Assert.AreEqual(128, CryptoUtils.Sha512Hash("hello").Length)
            Assert.AreEqual("", CryptoUtils.Sha512Hash(Nothing))
        End Sub

        ' =====================================================================
        ' 文件哈希测试
        ' =====================================================================

        <TestMethod>
        Public Sub TestSha256File()
            ' 创建临时文件
            Dim tempFile As String = Path.GetTempFileName()
            File.WriteAllText(tempFile, "test content", Encoding.UTF8)
            
            Try
                Dim hash As String = CryptoUtils.Sha256File(tempFile)
                Assert.AreEqual(64, hash.Length)
                Assert.IsTrue(CryptoUtils.IsValidSha256(hash))
                
                ' 相同内容应产生相同哈希
                Dim hash2 As String = CryptoUtils.Sha256File(tempFile)
                Assert.AreEqual(hash, hash2)
            Finally
                File.Delete(tempFile)
            End Try
        End Sub

        <TestMethod>
        Public Sub TestSha256FileNotExists()
            Assert.AreEqual("", CryptoUtils.Sha256File("/nonexistent/file.txt"))
        End Sub

        ' =====================================================================
        ' HMAC 测试
        ' =====================================================================

        <TestMethod>
        Public Sub TestHmacSha256()
            Dim message As String = "hello"
            Dim secret As String = "secret_key"
            Dim hmac As String = CryptoUtils.HmacSha256(message, secret)
            
            Assert.AreEqual(64, hmac.Length)
            Assert.IsTrue(CryptoUtils.IsValidSha256(hmac))
            
            ' 相同输入应产生相同输出
            Dim hmac2 As String = CryptoUtils.HmacSha256(message, secret)
            Assert.AreEqual(hmac, hmac2)
            
            ' 不同密钥应产生不同输出
            Dim hmac3 As String = CryptoUtils.HmacSha256(message, "different_key")
            Assert.AreNotEqual(hmac, hmac3)
        End Sub

        <TestMethod>
        Public Sub TestHmacSha256Empty()
            Assert.AreEqual("", CryptoUtils.HmacSha256(Nothing, "key"))
            Assert.AreEqual("", CryptoUtils.HmacSha256("message", Nothing))
            Assert.AreEqual("", CryptoUtils.HmacSha256("", ""))
        End Sub

        ' =====================================================================
        ' Base64 编码/解码测试
        ' =====================================================================

        <TestMethod>
        Public Sub TestBase64Encode()
            Assert.AreEqual("aGVsbG8=", CryptoUtils.Base64Encode("hello"))
            Assert.AreEqual("", CryptoUtils.Base64Encode(Nothing))
            Assert.AreEqual("", CryptoUtils.Base64Encode(""))
            
            ' 中文编码
            Dim encoded As String = CryptoUtils.Base64Encode("你好")
            Assert.IsTrue(encoded.Length > 0)
        End Sub

        <TestMethod>
        Public Sub TestBase64Decode()
            Assert.AreEqual("hello", CryptoUtils.Base64Decode("aGVsbG8="))
            Assert.AreEqual("", CryptoUtils.Base64Decode(Nothing))
            Assert.AreEqual("", CryptoUtils.Base64Decode(""))
            
            ' 无效输入
            Assert.AreEqual("", CryptoUtils.Base64Decode("invalid!!!"))
        End Sub

        <TestMethod>
        Public Sub TestBase64RoundTrip()
            Dim original As String = "Hello, 世界! 123 @#$"
            Dim encoded As String = CryptoUtils.Base64Encode(original)
            Dim decoded As String = CryptoUtils.Base64Decode(encoded)
            Assert.AreEqual(original, decoded)
        End Sub

        ' =====================================================================
        ' URL 安全 Base64 测试
        ' =====================================================================

        <TestMethod>
        Public Sub TestBase64UrlEncode()
            ' 包含 + 和 / 的字符串
            Dim input As String = ">>>???"
            Dim encoded As String = CryptoUtils.Base64UrlEncode(input)
            
            ' URL 安全编码不应包含 + / =
            Assert.IsFalse(encoded.Contains("+"))
            Assert.IsFalse(encoded.Contains("/"))
            Assert.IsFalse(encoded.Contains("="))
        End Sub

        <TestMethod>
        Public Sub TestBase64UrlRoundTrip()
            Dim original As String = "Hello+World/Test=Value"
            Dim encoded As String = CryptoUtils.Base64UrlEncode(original)
            Dim decoded As String = CryptoUtils.Base64UrlDecode(encoded)
            Assert.AreEqual(original, decoded)
        End Sub

        <TestMethod>
        Public Sub TestBase64UrlDecodeInvalid()
            Assert.AreEqual("", CryptoUtils.Base64UrlDecode("invalid"))
            Assert.AreEqual("", CryptoUtils.Base64UrlDecode(Nothing))
        End Sub

        ' =====================================================================
        ' 随机字符串测试
        ' =====================================================================

        <TestMethod>
        Public Sub TestRandomString()
            Dim s1 As String = CryptoUtils.RandomString(10)
            Assert.AreEqual(10, s1.Length)
            
            ' 两次生成的字符串应不同（概率极低相同）
            Dim s2 As String = CryptoUtils.RandomString(10)
            Assert.AreEqual(10, s2.Length)
            ' 注意：理论上可能相同，但概率极低
            
            ' 边界值
            Assert.AreEqual("", CryptoUtils.RandomString(0))
            Assert.AreEqual("", CryptoUtils.RandomString(-1))
        End Sub

        <TestMethod>
        Public Sub TestRandomStringCustomChars()
            Dim s As String = CryptoUtils.RandomString(20, "ABC")
            Assert.AreEqual(20, s.Length)
            ' 应只包含指定字符
            For Each c As Char In s
                Assert.IsTrue("ABC".Contains(c))
            Next
        End Sub

        <TestMethod>
        Public Sub TestRandomPassword()
            Dim password As String = CryptoUtils.RandomPassword(12)
            Assert.AreEqual(12, password.Length)
            
            ' 应包含至少一个大写字母
            Assert.IsTrue(password.Any(Function(c) Char.IsUpper(c)))
            ' 应包含至少一个小写字母
            Assert.IsTrue(password.Any(Function(c) Char.IsLower(c)))
            ' 应包含至少一个数字
            Assert.IsTrue(password.Any(Function(c) Char.IsDigit(c)))
            ' 应包含至少一个特殊字符
            Assert.IsTrue(password.Any(Function(c) "!@#$%^&*()_+-=[]{}|;:,.<>?".Contains(c)))
        End Sub

        <TestMethod>
        Public Sub TestRandomPasswordMinLength()
            ' 小于4应调整为8
            Dim p As String = CryptoUtils.RandomPassword(2)
            Assert.AreEqual(8, p.Length)
        End Sub

        ' =====================================================================
        ' UUID 生成测试
        ' =====================================================================

        <TestMethod>
        Public Sub TestGenerateUuid()
            Dim uuid1 As String = CryptoUtils.GenerateUuid()
            Dim uuid2 As String = CryptoUtils.GenerateUuid()
            
            ' 标准 UUID 格式: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
            Assert.AreEqual(36, uuid1.Length)
            Assert.AreEqual(4, uuid1.Split("-"c).Length - 1)
            
            ' 两次生成的 UUID 应不同
            Assert.AreNotEqual(uuid1, uuid2)
        End Sub

        <TestMethod>
        Public Sub TestGenerateUuidSimple()
            Dim uuid As String = CryptoUtils.GenerateUuidSimple()
            Assert.AreEqual(32, uuid.Length)
            Assert.IsFalse(uuid.Contains("-"))
        End Sub

        ' =====================================================================
        ' XOR 加密测试
        ' =====================================================================

        <TestMethod>
        Public Sub TestXorEncryptDecrypt()
            Dim original As String = "Hello, World!"
            Dim key As String = "my_secret_key"
            
            Dim encrypted As String = CryptoUtils.XorEncrypt(original, key)
            Assert.AreNotEqual(original, encrypted)
            
            Dim decrypted As String = CryptoUtils.XorDecrypt(encrypted, key)
            Assert.AreEqual(original, decrypted)
        End Sub

        <TestMethod>
        Public Sub TestXorEncryptEmpty()
            Assert.AreEqual("", CryptoUtils.XorEncrypt(Nothing, "key"))
            Assert.AreEqual("", CryptoUtils.XorEncrypt("text", Nothing))
            Assert.AreEqual("", CryptoUtils.XorDecrypt(Nothing, "key"))
            Assert.AreEqual("", CryptoUtils.XorDecrypt("text", Nothing))
        End Sub

        <TestMethod>
        Public Sub TestXorEncryptDifferentKeys()
            Dim original As String = "test message"
            Dim encrypted1 As String = CryptoUtils.XorEncrypt(original, "key1")
            Dim encrypted2 As String = CryptoUtils.XorEncrypt(original, "key2")
            
            ' 不同密钥应产生不同密文
            Assert.AreNotEqual(encrypted1, encrypted2)
            
            ' 用错误密钥解密应产生乱码（非原始值）
            Dim wrongDecrypt As String = CryptoUtils.XorDecrypt(encrypted1, "key2")
            Assert.AreNotEqual(original, wrongDecrypt)
        End Sub

        ' =====================================================================
        ' 混淆测试
        ' =====================================================================

        <TestMethod>
        Public Sub TestObfuscateDeobfuscate()
            Dim original As String = "Secret Message 123!"
            Dim obfuscated As String = CryptoUtils.Obfuscate(original)
            
            Assert.AreNotEqual(original, obfuscated)
            
            Dim deobfuscated As String = CryptoUtils.Deobfuscate(obfuscated)
            Assert.AreEqual(original, deobfuscated)
        End Sub

        <TestMethod>
        Public Sub TestObfuscateEmpty()
            Assert.AreEqual("", CryptoUtils.Obfuscate(Nothing))
            Assert.AreEqual("", CryptoUtils.Obfuscate(""))
            Assert.AreEqual("", CryptoUtils.Deobfuscate(Nothing))
            Assert.AreEqual("", CryptoUtils.Deobfuscate(""))
        End Sub

        <TestMethod>
        Public Sub TestObfuscateInvalid()
            ' 无效输入应返回空字符串
            Assert.AreEqual("", CryptoUtils.Deobfuscate("not_valid_base64!!!"))
        End Sub

        ' =====================================================================
        ' 验证函数测试
        ' =====================================================================

        <TestMethod>
        Public Sub TestIsValidMd5()
            Assert.IsTrue(CryptoUtils.IsValidMd5("5d41402abc4b2a76b9719d911017c592"))
            Assert.IsTrue(CryptoUtils.IsValidMd5("D41D8CD98F00B204E9800998ECF8427E")) ' 大写
            Assert.IsFalse(CryptoUtils.IsValidMd5("invalid"))
            Assert.IsFalse(CryptoUtils.IsValidMd5(""))
            Assert.IsFalse(CryptoUtils.IsValidMd5(Nothing))
            Assert.IsFalse(CryptoUtils.IsValidMd5("5d41402abc4b2a76b9719d911017c59")) ' 31位
            Assert.IsFalse(CryptoUtils.IsValidMd5("5d41402abc4b2a76b9719d911017c5923")) ' 33位
        End Sub

        <TestMethod>
        Public Sub TestIsValidSha256()
            Assert.IsTrue(CryptoUtils.IsValidSha256("2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824"))
            Assert.IsFalse(CryptoUtils.IsValidSha256("invalid"))
            Assert.IsFalse(CryptoUtils.IsValidSha256(""))
            Assert.IsFalse(CryptoUtils.IsValidSha256(Nothing))
            Assert.IsFalse(CryptoUtils.IsValidSha256("abc123")) ' 太短
        End Sub

        <TestMethod>
        Public Sub TestIsValidBase64()
            Assert.IsTrue(CryptoUtils.IsValidBase64("aGVsbG8="))
            Assert.IsTrue(CryptoUtils.IsValidBase64("dGVzdA=="))
            Assert.IsTrue(CryptoUtils.IsValidBase64(""))
            Assert.IsFalse(CryptoUtils.IsValidBase64(Nothing))
            Assert.IsFalse(CryptoUtils.IsValidBase64("invalid!!!"))
            Assert.IsFalse(CryptoUtils.IsValidBase64("aGVsbG8")) ' 缺少填充但可能有效
        End Sub

        ' =====================================================================
        ' 性能/大数据测试
        ' =====================================================================

        <TestMethod>
        Public Sub TestLargeDataHashing()
            Dim largeString As String = New String("a"c, 1000000) ' 1MB 数据
            Dim hash As String = CryptoUtils.Sha256Hash(largeString)
            Assert.AreEqual(64, hash.Length)
            Assert.IsTrue(CryptoUtils.IsValidSha256(hash))
        End Sub

        <TestMethod>
        Public Sub TestUnicodeSupport()
            Dim unicodeStrings As String() = New String() {
                "Hello 世界",
                "🎉 Emoji test 🚀",
                "العربية",
                "עברית",
                "日本語テキスト"
            }
            
            For Each s As String In unicodeStrings
                Dim hash As String = CryptoUtils.Md5Hash(s)
                Assert.AreEqual(32, hash.Length, "Failed for: " & s)
                
                Dim encoded As String = CryptoUtils.Base64Encode(s)
                Dim decoded As String = CryptoUtils.Base64Decode(encoded)
                Assert.AreEqual(s, decoded, "Base64 round-trip failed for: " & s)
            Next
        End Sub

    End Class

End Namespace
